plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+                        if digest_value and digest_value == getattr(artifact, digest_name):
+                            declarative_artifact.artifact = artifact
+        for content in declarative_content:
+            if content is None:
+                continue
+            await out_q.put(content)
+        declarative_content = []
+        if shutdown:
+            break
+    await out_q.put(None)
+
+
+class artifact_downloader:
+    """
+    An object containing a Stages API stage to download :class:`~pulpcore.plugin.models.Artifact`
+    file, but don't save the :class:`~pulpcore.plugin.models.Artifact` in the db.
+
+    The actual stage is the :meth:`~pulpcore.plugin.stages.artifact_downloader.stage`
+    which can be used as follows:
+
+    >>> artifact_downloader(max_concurrent_downloads=42).stage  # This is the real stage
+
+    This stage downloads the file for any :class:`~pulpcore.plugin.models.Artifact` objects missing
+    files and creates a new :class:`~pulpcore.plugin.models.Artifact` object from the downloaded
+    file and its digest data. The new :class:`~pulpcore.plugin.models.Artifact` is not saved but
+    added to the :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object, replacing the likely
+    incomplete :class:`~pulpcore.plugin.models.Artifact`.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage creates a ProgressBar named 'Downloading Artifacts' that counts the number of
+    downloads completed. Since it's a stream the total count isn't known until it's finished.
+
+    This stage drains all available items from `in_q` and starts as many downloaders as possible.
+
+    Args:
+        max_concurrent_downloads (int): The maximum number of concurrent downloads this stage will
+            run. Default is 100.
+
+    Returns:
+        An object containing the artifact_downloader stage to be included in a pipeline.
+    """
+
+    def __init__(self, max_concurrent_downloads=100):
+        self.max_concurrent_downloads = max_concurrent_downloads
+
+    async def stage(self, in_q, out_q):
+        """
+        Download undownloaded Artifacts, but don't save them in the db.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from that may have
+                undownloaded files.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects into, all of which have
+                files downloaded.
+
+        """
+        pending = set()
+        incoming_content = []
+        outstanding_downloads = 0
+        shutdown = False
+        saturated = False
+        with ProgressBar(message='Downloading Artifacts') as pb:
+            while True:
+                if not saturated:
```

Comment:
This *batching* logic seems to rely on a momentary gap in the `in_q` to determine the batch size.  If the `in_q` is being fed faster than consumed here, won't that result in all of the content being in memory at once.  Conversely, what if the `in_q` is being fed in slow, small bursts.  Won't that result in really small batch sizes and poor performance?

The `saturated` flag is set true later in the loop so don't think it will prevent this.

Reply:
re batching logic: we should tune the stages later with data. I have a similar answer [here](https://github.com/pulp/pulp/pull/3559#discussion_r206942577).

The saturated feature is for limiting the number of downloaders running. It's possible that a greater number of things to download could be drained from the queue already at which point on the next loop run nothing is read from the queue because it's saturated.